### ISSUE TITLE
CSharpCompiler makes rust server semi-freeze

### ISSUE THREAD

**Issue by neico** [OxideMod/Oxide#477]
If you reload a C# plugin it usually starts up CSharpCompiler and after compiling it loads the assembly.
But the compiler keeps running (around 50MB wasted memory), but no CPU usage.

If you now reload for example a lua plugin however it'll most likely freeze the rust server for a while and the compiler is taking up at least 30% CPU on my server, after killing the compiler it restarts the compiler and everything goes fine.

I think I also had the case where it simply reloaded the lua plugin and the compiler vanished, seems very random (didn't had that much time to look at it exactly, was still developing)

In any case this is a rather serious issue since Oxide or any of it's tools or plugins should ever generate possible downtimes on the game server, my guess is that it's not handling the reload mechanism correctly and doesn't compile C# plugins fully asynchron (sending the plugin of for compilation and going on and once the compiler is done it simply loads it, without waiting for the compiler to finish first)

But as I said, I only had briefly time to look into it, not to mention that I'm by far not very experienced with C# in any way, so there's that...

**Comment by lukespragg** [comment]
The compiler should automatically shutdown after 30 seconds of inactivity. If you do not see this, it's possible that you are using an outdated build.

**Comment by neico** [comment]
Oxide and the patcher are always being downloaded directly from the latest snapshot each time I reboot the server: http://puu.sh/jkKPR/3f1bb3c2e3.txt

So it's pretty much the most recent one

And not sure about the 30 seconds, but apparently it happens at least within those 30 seconds, which can totally happen if you just drag&drop the new files in there

**Comment by neico** [comment]
I seem to have found a edge case with the compiler: the compiler kills itself after 1 minute, before it does that it's raising it's CPU usage to 30% and then kills itself off, whatever it does in that time seems to critical that if you happen to send a plugin to compile in that moment the compiler will get stuck.

It won't compile anything for at least 1 minute (unless you kill the process, in which case oxide starts a new one which compiles fine then), since I only waited so long once I can't really remember what happened then but I think it compiled the file and then shutdown, or it shutdown and oxide spawned a new one like above...?

Anyway, seems that the compiler itself seems at fault here...

**Comment by lukespragg** [comment]
We'll look into it and see if we can reproduce it based on what you've given us. :)

**Comment by lukespragg** [comment]
I've been testing trying to reproduce this, but so far I've have been able to. The CPU usage for CSharpCompiler.exe never spikes above 2% usage with my 8 logical cores. No plugins hang during compiling, reloads have been triggered throughout the time while the compiler was running up until the point where it shuts down automatically from being idle.

**Comment by neico** [comment]
Well that's strange, I can't remember when it started but I noticed at some point that booting up the rust server takes a long time, when I noticed this issue I also noticed that the compiler seems to hang up somewhere when the rust server is loading the map elements like icebergs etc.

If I kill the process before that then the server boots up fine, so it must be something that the compiler does wrong after a while. Guess I'll have to debug the process myself to see what's happening exactly...

**Comment by lukespragg** [comment]
The Rust server actually does a lot during startup, backup up your entire identity folder, deleting old signage, etc. The delay isn't related to Oxide generally. All of the other servers that Oxide supports are much faster to startup.

From what @bawNg told me:
```
The compiler blocks boot up, long before procgen starts generating anything.

So by the time the server starts printing stuff to console, the compiler is already done.
```

**Comment by neico** [comment]
Seems like the latest changes have fixed this, I'll observe if it appears again for a bit longer before closing this...

**Comment by gs9511** [comment]
I am running in a similar issue, the server loads all plugins and then hangs for around 5 minutes waiting for the CSharpCompiler.exe executable to finish.
This issue only happens when a .cs plugin is loaded(I tested multiple ones).

I am running the latest build (Oxide 2.0.1333 for Protocol 1312) on Windows 2012 R2

**Comment by lukespragg** [comment]
Please try the latest Oxide snapshots and let us know if you still have any issues.

**Comment by neico** [comment]
There were several updates (be it Oxide or the compiler) that fixed the issue, but the most recent version shows the same issue again...

And it has nothing to do with plugins failing to compile or so, it happens when the compiler is supposed to shutdown, but instead locks up and with it Oxide / the game server...

I get the feeling that the injection method used to patch the compiler seems to be unstable and sometimes produces the correct results and sometimes don't...

**Comment by lukespragg** [comment]
The compiler isn't being injected into anything, it's mkbundled modified Mono basic.exe compiler. It hasn't changed between the versions you've used really, the only thing that has changed is Oxide's handling of the compiled plugins to fix various issues, bugs, etc.

The timeout still exists, so the compiler shouldn't be keeping the server from starting being that period. I'm not sure why you're having the issue, but we'd need to be able to reproduce the issue, which is hard as you're currently the only user experiencing the issue. @gs9511's issue is a bit different, but should already be resolved.

**Comment by neico** [comment]
The compiler started with his usual 30% CPU usage (locking up the game server) right after these two lines from it's log:

```
[SERVER v1.0.$env:APPVEYOR_BUILD_NUMBER] Got Message: Exit
[SERVER v1.0.$env:APPVEYOR_BUILD_NUMBER] Exit received. Shutdown
```
(Adding the time of the log message could probably improve on debugging as well)

So it must be something that keeps the compiler running against it's will...

If I only had a pdb or so that would allow me to see what the compiler is trying to do, then this would be easier to hunt down, I can't really think of any reason why it might do that.

I'm starting to get the hang of where is which code stored...

first of, this compiler service code has quite the horrible indentation...
secondly, https://github.com/OxideMod/Mono/blob/master/mcs/mcs/driver.cs#L335 seems fishy, like you're killing the app right of without making sure that the event loop, thread etc. are made aware of that the app is going to close now, no GC cleanup, nothing...

**Comment by lukespragg** [comment]
The indentation was borked a bit by @Nogrod's IDE when he merged from upstream Mono. He did update our fork and made a change, so I'd recommend trying the latest Oxide snapshot when you're able to see if it is still an issue.

**Comment by neico** [comment]
Perfect, that did the trick, working totally as expected now :D 

```
[SERVER v1.0.18] Got Message: Exit
[SERVER v1.0.18] Exit received.
[SERVER v1.0.18] Connection closed.
[SERVER v1.0.18] Shutdown
```

I don't expect this to happen anymore but will continue to observe it over the days, just to be sure.
